source/common/common/key_value_store_base.cc

```diff
@@ -63,8 +65,10 @@ bool KeyValueStoreBase::parseContents(absl::string_view contents,
 }
 
 void KeyValueStoreBase::addOrUpdate(absl::string_view key, absl::string_view value) {
-  store_.erase(key);
-  store_.emplace(key, value);
+  // Avoids rehashing if the key already exists, moving the value into the map.
+  // If the key did not already exists, inserts both into the map, which
+  // might trigger a rehash.
+  store_.insert_or_assign(key, std::string(value));
```

Comment:
This idiom (insert_or_assign) is shorter than erase();emplace(); so that sounds good to me. But I don't think I understand why this code needs to avoid causing a rehash here? Is this a performance thing, or correctness?